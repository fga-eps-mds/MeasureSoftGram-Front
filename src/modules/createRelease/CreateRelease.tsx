import React, { useEffect, useState } from 'react';
import { Alert, Container, Snackbar } from '@mui/material';

import DrawerMenu from '@components/DrawerMenu';
import { ButtonType, Product } from '@customTypes/product';
import ConfigPage from '@modules/preConfig/ConfigPage';
import ReleaseInfo from './components/ReleaseInfo';
import ReleaseGoals from './components/ReleaseGoals';

import { CreateReleaseProvider, useCreateReleaseContext } from './context/useCreateRelease';
import FirstReleaseWarning from './components/FirstReleaseWarning';
import ReleaseConfigSelector from './components/ReleaseConfigSelector/ReleaseConfigSelector';
import ThresholdConfig from './components/ThresholdConfig/ThresholdConfig';
import { CREATE_RELEASE_STEP } from './consts';

interface CreateReleaseProps {
  open: boolean;
  handleClose: () => void;
}

interface CreateReleaseContainerProps {
  productId: string;
  organizationId: string;
  currentProduct: Product;
  open: boolean;
  handleClose: () => void;
}

function CreateRelease({ open, handleClose }: CreateReleaseProps) {
  const [activeStep, setActiveStep] = useState(0);
  const [configPage, setConfigPage] = useState(0);
  const [configPageTitle, setConfigPageTitle] = useState<string>('');
  const configPageTitles = [
    'Definir peso das características',
    'Definir peso das subcaracterísticas',
    'Definir peso das medidas'
  ];

  const {
    successOnCreation,
    closeAlert,
    goToNextStep,
    createProductReleaseGoal,
    organizationId,
    productId,
    releaseInfoForm,
    configPageData,
    getNextStep,
    getPreviousStep,
    isFirstRelease
  } = useCreateReleaseContext();

  const renderStep = () =>
  ({
    0: <ReleaseInfo />,
    1: isFirstRelease ? <FirstReleaseWarning /> : <ReleaseConfigSelector setActiveStep={setActiveStep} />,
    2: (
      <ConfigPage
        page={configPage}
        isOpen
        onClose={handleClose}
        organizationId={organizationId}
        filteredCharacteristics={releaseInfoForm?.characteristics}
        productId={productId}
        title={configPageTitle}
      />
    ),
    3: (
      <ThresholdConfig
        onChange={() => { }}
        data={configPageData.characteristicData}
        tabs={configPageData.subcharacterCheckbox}
        checkboxValues={configPageData.measureCheckbox}
        setCheckboxValues={configPageData.setMeasureCheckbox}
      />
    ),
    4: <ReleaseGoals />
  }[activeStep]);

  const handleCloseModal = () => {
    handleClose();
    closeAlert();
  };

  const handleGoToNextStep = () => {
    if (!goToNextStep()) return;

    const newActiveStep = getNextStep(activeStep, configPage);
    if (activeStep === newActiveStep && newActiveStep === CREATE_RELEASE_STEP.CharacteristicStep) {
      setConfigPage(configPage + 1);
    }
    setActiveStep(newActiveStep);
  };

  const handleBackButton = () => {
    if (activeStep === CREATE_RELEASE_STEP.ReleaseInfoStep) {
      handleCloseModal();
      return;
    }

    const newActiveStep = getPreviousStep(activeStep, configPage);
    if (activeStep === newActiveStep && newActiveStep === CREATE_RELEASE_STEP.CharacteristicStep) {
      setConfigPage(configPage - 1);
    }
    setActiveStep(newActiveStep);
  };

  useEffect(() => {
    setConfigPageTitle(configPageTitles[configPage]);
  }, [configPage]);

  const handleNextButton = () =>
    activeStep === CREATE_RELEASE_STEP.ReleaseGoalStep ? createProductReleaseGoal() : handleGoToNextStep();

  const BUTTONS: Array<ButtonType> = [
    {
      label: 'Voltar',
      onClick: handleBackButton,
      backgroundColor: '#FFF',
      color: '#113D4C',
      variant: 'outlined',
    },
    {
      label: activeStep === CREATE_RELEASE_STEP.ReleaseGoalStep ? 'Finalizar' : 'Continuar',
      onClick: handleNextButton,
      backgroundColor: '#113D4C',
      color: '#fff',
      variant: 'contained',
      disabled: activeStep === CREATE_RELEASE_STEP.ReleaseSelectorStep && !isFirstRelease
    }
  ];

  const ALERTS = {
    success: <Alert severity="success">Release criada com sucesso!</Alert>,
    error: <Alert severity="error">Ocorreu um erro durante a criação da release.</Alert>
  };
  const handleOnCloseAlert = () => {
    if (successOnCreation === 'success') {
      handleClose();
      setActiveStep(0);
    }

    setTimeout(closeAlert, 3000);
  };

  useEffect(() => {
    if (successOnCreation === 'success') handleOnCloseAlert();
  }, [successOnCreation]);

  return (
    <>
      <DrawerMenu open={open} buttons={BUTTONS}>
        <Container>{renderStep()}</Container>
      </DrawerMenu>

      <Snackbar
        anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        open={!!successOnCreation}
        onClose={handleOnCloseAlert}
      >
        {ALERTS[successOnCreation as keyof typeof ALERTS]}
      </Snackbar>
    </>
  );
}

function CreateReleaseContainer({
  productId,
  organizationId,
  open,
  handleClose,
  currentProduct
}: CreateReleaseContainerProps) {
  return (
    <CreateReleaseProvider productId={productId} organizationId={organizationId} currentProduct={currentProduct}>
      <CreateRelease open={open} handleClose={handleClose} />
    </CreateReleaseProvider>
  );
}

export default CreateReleaseContainer;
