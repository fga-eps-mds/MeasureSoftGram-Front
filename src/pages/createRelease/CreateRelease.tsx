import React, { useState } from 'react'
import { Alert, Container, Snackbar } from '@mui/material';

import DrawerMenu from '@components/DrawerMenu';
import { ButtonType } from '@customTypes/project';
import ReleaseInfo from './components/ReleaseInfo';
import ReleaseGoals from './components/ReleaseGoals';

import { CreateReleaseProvider, useCreateReleaseContext } from './context/useCreateRelease';

interface CreateReleaseProps {
  open: boolean;
  handleClose: () => void;
}

interface CreateReleaseContainerProps {
  projectId: number;
  organizationId: number;
  open: boolean;
  handleClose: () => void;
}

function CreateRelease({ open, handleClose }: CreateReleaseProps) {
  const [activeStep, setActiveStep] = useState(0);

  const {
    successOnCreation,
    closeAlert,
    goToGoalsStep,
    createProjectReleaseGoal,
  } = useCreateReleaseContext();

  const renderStep = () => ({
      0: <ReleaseInfo/>,
      1: <ReleaseGoals />
    }[activeStep])

  const handleCloseModal = () => {
    handleClose()
    closeAlert()
  }

  const handleGoToGoalsStep = () => goToGoalsStep() ? setActiveStep(1) : () => {}

  const handleBackButton = () => activeStep ? setActiveStep(0) : handleCloseModal();

  const handleNextButton = () => activeStep ? createProjectReleaseGoal() : handleGoToGoalsStep();

  const BUTTONS: Array<ButtonType> = [
    {
      label: 'Voltar',
      onClick: handleBackButton,
      backgroundColor: '#FFF',
      color: '#113D4C',
      variant: 'outlined'
    },
    {
      label: activeStep ? 'Finalizar' : 'Continuar',
      onClick: handleNextButton,
      backgroundColor: '#113D4C',
      color: '#fff',
      variant: 'outlined'
    }
  ];

  const ALERTS = {
    success: (
      <Alert severity="success">Release criada com sucesso!</Alert>
    ),
    error: (
      <Alert severity="error">Ocorreu um erro durante a criação da release.</Alert>
    )
  }
  const handleOnCloseAlert = () => {
    if (successOnCreation === 'success') {
      handleClose()
      setActiveStep(0)
    }

    closeAlert()
  }

  return (
    <>
      <DrawerMenu open={open} buttons={BUTTONS}>
        <Container>
          {renderStep()}
        </Container>
      </DrawerMenu>

      <Snackbar
        anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        open={!!successOnCreation}
        autoHideDuration={3000}
        onClose={handleOnCloseAlert}
      >
        {ALERTS[successOnCreation as keyof typeof ALERTS]}
      </Snackbar>
    </>
  )
}

function CreateReleaseContainer({
  projectId,
  organizationId,
  open,
  handleClose
}: CreateReleaseContainerProps) {
  return (
    <CreateReleaseProvider
      projectId={projectId.toString()}
      organizationId={organizationId.toString()}
    >
      <CreateRelease
        open={open}
        handleClose={handleClose}
      />
    </CreateReleaseProvider>
  )
}

export default CreateReleaseContainer;
