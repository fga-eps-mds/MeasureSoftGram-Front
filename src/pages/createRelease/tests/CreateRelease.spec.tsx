import '@testing-library/jest-dom';

import React from 'react';
import { render } from '@testing-library/react';

import CreateRelease from '../CreateRelease';

describe('<CreateRelease />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <CreateRelease
          open
          handleClose={() => {}}
          productId={1}
          organizationId={1}
        />
      )

      expect(tree).toMatchSnapshot();
    });
  });
});
