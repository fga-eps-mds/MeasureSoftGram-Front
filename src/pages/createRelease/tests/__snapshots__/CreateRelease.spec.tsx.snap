// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateRelease /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDrawer-root MuiDrawer-modal MuiModal-root css-1u2w381-MuiModal-root-MuiDrawer-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorRight css-1160xiw-MuiPaper-root-MuiDrawer-paper"
        tabindex="-1"
      >
        <div
          class="MuiBox-root css-1m45ylu"
        >
          <div
            class="styles__ScrollDiv-sc-1casz44-0 eTyGvR"
          >
            <div
              class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
            >
              <div
                class="MuiBox-root css-0"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                  style="width: 300px; height: 64px;"
                />
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                  style="width: 600px; height: 32px;"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-1yuhvjn"
          >
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1mxmpgj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Voltar
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1bjby6n-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Continuar
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
